jobs/gtfs-schedule-validator/gtfs_schedule_validator.py

```diff
@@ -186,90 +265,77 @@ def validate_day(
-
-            typer.secho(
-                f"saving {len(notices)} validation notices to {validation.path}",
-                fg=typer.colors.GREEN,
-            )
-            validation.save_content(
-                content=gzip.compress(
-                    "\n".join(json.dumps(notice) for notice in notices).encode()
-                ),
-                fs=fs,
-            )
-            outcomes.append(
-                GTFSScheduleFeedExtractValidationOutcome(
-                    success=True,
-                    extract_path=extract.path,
-                    validation=validation,
-                )
-            )
-        except Exception as e:
-            typer.secho(
-                f"encountered exception on extract {extract.path}: {e}\n{traceback.format_exc()}",
-                fg=typer.colors.RED,
-            )
-            if verbose and isinstance(e, subprocess.CalledProcessError):
-                typer.secho(
-                    e.stderr,
+    with ThreadPoolExecutor(max_workers=threads) as pool:
+        futures: Dict[Future, GTFSScheduleFeedExtract] = {
+            pool.submit(
+                download_and_validate_extract,
+                extract=extract,
+                pbar=pbar,
+            ): extract
+            for i, extract in enumerate(extracts)
+        }
+
+        for future in concurrent.futures.as_completed(futures):
+            extract = futures[future]
+            if pbar:
+                pbar.update(1)
+
+            try:
+                outcomes.append(future.result())
+            except KeyboardInterrupt:
+                raise
+            except Exception as e:
+                log(
+                    f"encountered exception on extract {extract.path}: {e}\n{traceback.format_exc()}",
                     fg=typer.colors.RED,
+                    pbar=pbar,
                 )
-            outcomes.append(
-                GTFSScheduleFeedExtractValidationOutcome(
-                    success=False,
-                    extract_path=extract.path,
-                    exception=e,
+                if verbose and isinstance(e, subprocess.CalledProcessError):
+                    log(
+                        e.stderr,
+                        fg=typer.colors.RED,
+                        pbar=pbar,
+                    )
+
+                exceptions.append(e)
+                outcomes.append(
+                    GTFSScheduleFeedExtractValidationOutcome(
+                        success=False,
+                        extract=extract,
+                        exception=e,
+                    )
                 )
-            )
+
+    if pbar:
+        del pbar
+
     result = ScheduleValidationJobResult(
         filename="results.jsonl",
         dt=day,
         outcomes=outcomes,
     )
     typer.secho(
-        f"got {len(result.successes)} successes and {len(result.failures)} failures",
+        f"got {len(result.successes)} successes and {len(exceptions)} failures",
```

Comment:
why use exceptions rather than `result.failures` here?

Reply:
no real reason, they should be the same; I'll add an assert for that

Reply:
mostly I've always kinda been 🤷 about the utility of the `results.<type>` methods and if they're there I feel like we should use them....?